README.md
## `enterprise-components` for `kdb+`

`enterprise-components` are a set of open source building blocks on top of `kdb+` allowing you to setup a system for data warehousing and 
analytics in just a few simple steps. It facilitates building of the professional data system covering:
- data import
- storage
- quality assurance
- data analytics
- data extraction
- interfacing with other systems

User can fully customize system architecture through configuration files without `q` knowledge. 

![Overview](doc/img/Slide1.PNG) 


### Download

Most recent `DemoSystem` can be downloaded from the [releases page](https://github.com/exxeleron/enterprise-components/releases).

### Documentation
Some basic information can be found below, for more extensive documentation refer to the [documentation folder](doc/README.md).

For list of features, changes and fixes in each release refer to [changelog](CHANGELOG.txt) or 
[releases page](https://github.com/exxeleron/enterprise-components/releases).

### Tutorial
Beginners to `kdb+`/`q` and `enterprise-components` should probably start with the [tutorial](tutorial) that covers the essential basics 
via set of lessons.

## General description

### Objective
`enterprise-components` have been designed with the goal to provide core functionality to `kdb+` installations:
- complying with well-known IT policies for enterprise infrastructures
- keeping the performance and flexibility of `kdb+`
- making `kdb+` development easier as well as faster and infrastructures more robust
- ensuring efficient operations of `kdb+` infrastructures

![Overview](doc/img/Slide2.PNG)

In order to achieve these goals Exxeleron’s solutions follow best practices for robust and efficient IT service delivery and system 
architecture. `enterprise-components` development is guided by overall TCO (total cost of ownership) considerations where special attention 
is given to: 
- System Vision - new infrastructure should fit naturally into the architectural blueprint, ensuring long-term return on investment 
- Flexibility and Low Coupling - avoiding dependencies and complexity where possible 
- Reuse and Modularity - profiting from investments within their respective core capabilities 
- Flexibility - adaptation to changing business needs should be possible with small investments
- Service Orientation - low TCO and well behaved integration functionality silos 
- Low Maintenance - robustness and automation built-in 

Customers building their `kdb+` architectures with `enterprise-components` benefit from:
- reduced time-to-market cycles due to the wide set of supported features available out-of-the-box
- efficiently managed operational risks due to proven `kdb+` design principles
- simpler architectures due to consistent system design and system operations
- motivated `kdb+` teams focusing on core business requirements rather than basic infrastructure functionality

### Key features
`enterprise-components` provide an optimized set of building blocks that can get quickly assembled into complex `kdb+` deployments. 

![Overview](doc/img/Slide3.PNG)

Building blocks are based on common foundation - set of libraries built on top of `q`. The libraries are unifying and facilitating usage of 
various `q` resources. New building blocks can be easily created by using these libraries which enable development process and allow natural 
integration with the rest of the system.

![Overview](doc/img/Slide4.PNG)

`enterprise-components` enable management of the data flow at all levels, starting from high-frequency data capture and stream distribution 
through on-stream, in-memory, and on-disk data analysis to interface components for applications and programming languages.

![Overview](doc/img/Slide5.PNG)

`enterprise-components` functionality is driven by real-life needs. It combines DEVnet experience with modular  `kdb+` installations that 
harness `kdb+'s` full power and flexibility. 

![Overview](doc/img/Slide6.PNG)

With scalability, flexibility, robustness, optimization and consistent configuration as design goals it plays well in enterprise deployment.

![Overview](doc/img/Slide7.PNG)



CHANGELOG.txt
------------------------------------------------------------------------------
  enterprise-components 3.3.2 [2018.01.12]
------------------------------------------------------------------------------

  - Fix: refreshUFiles = remove spurious warnings (#53)
  - Fix: monitor - added a check if process name configured in dataflow.cfg is defined in system.cfg (#54)
  - Fix: hk - added a check for directory existence to log a warning instead of errror (#56)

------------------------------------------------------------------------------
  enterprise-components 3.3.1 [2016.06.03]
------------------------------------------------------------------------------

  - New: qsl/doc - new library - access to the API documentation from the q process level
  - New: qsl/sl - new wrapping function for garbage collector
  - Change: doc - major upgrade of the source code documentation
  - Change: qsl/doc - new global table .sl.libs
  - Fix: cfgRdr - now ? character allowed in STRING (#51)
  - Fix: qsl/sl - prevent housekeeping script to delete init.log (#52)
  
------------------------------------------------------------------------------
  enterprise-components 3.3.0 [2015.09.15]
------------------------------------------------------------------------------

  - New: hdbWriter - new component - writing data to the hdb
  - New: rproc - new component - realtime processing
  - New: qtest - ec test framework
  - New: mock/libs - set of predefined mocks
  - New: qsl/tabs - new library - data model validation utilities
  - New: tests - qtest-based tests for rdb, hdb, hdbWriter, qsl/tabs, qsl/handle
  - Change: doc - wiki content moved to main repo - doc directory and README.md files
  - Change: DemoSystem - packaging and installation instructions reworked
  - Fix: qsl/event monitor - fix processing events on Windows OS (#46)
  - Fix: authentication - add logic handling some misconfigurations in configuration files (#47)
  - Fix: hdb - corrected typo in .hdb.status[] in format column (#48)
 
 ------------------------------------------------------------------------------
  enterprise-components 3.2.3 [2015.02.11]
------------------------------------------------------------------------------

  - Fix: stream - port closed plugin fixed (#43)
  
------------------------------------------------------------------------------
  enterprise-components 3.2.2 [2015.01.14]
------------------------------------------------------------------------------

  - New: qsl/parseq - new parser combinator - `sequence` (#42)
  - Fix: rdb/replay - replaying data from journals fixed (#40)
  - Fix: cfgRdr - corner case of loading qsd files corrected (now eodMng works correctly with sync.cfg) (#41)

------------------------------------------------------------------------------
  enterprise-components 3.2.1 [2014.12.18]
------------------------------------------------------------------------------

  - New: stream/streamWdb – add new field performSort in streamWdb.q plugin (#32)
  - New: eodMng/hdbHk - add new field performBackup to enable/disable backup of hdb partition before housekeeping
  - New: monitor - proper handling not existing functions in sysFuncSummary statistics (#36)
  - Fix: feedMng - fix subscription using sub.q library (#27)
  - Fix: eodMng/hdbHk - fix compression plugin (#34)
  - Fix: qsl/os - fix a double cast bug in .os.find (#31)
  - Fix: rdb/replay - fix in starting replay.q (#32)

------------------------------------------------------------------------------
  enterprise-components 3.2 [2014.12.08]
------------------------------------------------------------------------------

  - New: general - experimental support for OS X (#17)
  - New: cfgRdr - support global custom configuration fields on the top level (#25)
  - New: qsl/os - .os.system[] - wrapper for the system calls with errors thrown as signal
  - New: qsl/os - .os.sleep[] - sleeps for given number of milliseconds
  - Fix: qsl/handle - now failure of `lazy connection is logged as warn (instead of error)
  - Fix: dist - .dist.usub[] - optimization of the un-subscription method
  
------------------------------------------------------------------------------
  enterprise-components 3.1 [2014.08.06]
------------------------------------------------------------------------------

  - New: general - experimental support for Windows (known limitations listed in Troubleshooting_windows) (#5)
  - New: qsl/os - new library with shell commands abstraction (#5)
  - New: dist - .dist.addTable[] - support for ad-hoc tables (#10)
  - New: cfgRdr - `commonLibs` configuration field for loading common q and/or qsd to a group of processes (#6)
  - New: tutorial - Lesson 5 - Authentication and authorization
  - Fix: monitor - [sysHdbSummary] and [sysDiskUsage] configuration handling (#8 and #9)
  - Fix: qsl/sl (Windows) - remove administrator privileges requirement (#22)

------------------------------------------------------------------------------
  enterprise-components 3.0 [2014.05.16]
------------------------------------------------------------------------------

  - Initial open source version
  - Components: accessPoint, dist, feedCsv, hdb, mock, rdb, tickHF, authentication, eodMng, feedMng, hk, monitor, stream, tickLF
  - Libraries: cfgRdr, qsl

------------------------------------------------------------------------------


test/runAllTests.sh
# --------------------------------------------------------------------------- #
#                       executing all qtest-style tests                       #
# --------------------------------------------------------------------------- #
if [ ! -d "bin" ]; then
  echo "Expecting bin subdirectory" && return
fi
export EC_ROOT_PATH=${PWD}

# ------ qsl/handle tests --------------------------------------------------- #
cd ${EC_ROOT_PATH}
source bin/ec/libraries/qsl/test/handle/etc/env.sh
yak console t.run -a " -quitAfterRun"

# ------ qsl/tabs tests ----------------------------------------------------- #
cd ${EC_ROOT_PATH}
source bin/ec/libraries/qsl/test/tabs/etc/env.sh
yak console t.run -a " -quitAfterRun"

# ------ rdb tests ---------------------------------------------------------- #
cd ${EC_ROOT_PATH}
source bin/ec/components/rdb/test/etc/env.sh
yak console t.run -a " -quitAfterRun"

# ------ hdb tests ---------------------------------------------------------- #
cd ${EC_ROOT_PATH}
source bin/ec/components/hdb/test/etc/env.sh
yak console t.run -a " -quitAfterRun"

# ------ hdbWriter tests ---------------------------------------------------- #
cd ${EC_ROOT_PATH}
source bin/ec/components/hdbWriter/test/etc/env.sh
yak console t.run -a " -quitAfterRun"

# ------ Lesson1 tests ---------------------------------------------------- #
cd ${EC_ROOT_PATH}
source bin/ec/test/tutorialTest/etc/env.sh
yak console t.run1 -a " -quitAfterRun"

# ------ Lesson2 tests ---------------------------------------------------- #
cd ${EC_ROOT_PATH}
source bin/ec/test/tutorialTest/etc/env.sh
yak console t.run2 -a " -quitAfterRun"

# ------ Lesson3 tests ---------------------------------------------------- #
cd ${EC_ROOT_PATH}
source bin/ec/test/tutorialTest/etc/env.sh
yak console t.run3 -a " -quitAfterRun"

# ------ Lesson4 tests ---------------------------------------------------- #
cd ${EC_ROOT_PATH}
source bin/ec/test/tutorialTest/etc/env.sh
yak console t.run4 -a " -quitAfterRun"

# ------ Lesson5 tests ---------------------------------------------------- #
cd ${EC_ROOT_PATH}
source bin/ec/test/tutorialTest/etc/env.sh
yak console t.run5 -a " -quitAfterRun"

# ------ collect test results ----------------------------------------------- #
cd ${EC_ROOT_PATH}
source bin/ec/test/collectTestResults/etc/env.sh
yak console t.collectResults

# --------------------------------------------------------------------------- #


test/tutorialTest/etc/env.sh
if [ ! -d "bin" ]; then
  echo "Expecting bin subdirectory" && return
fi
export EC_SYS_PATH=${PWD}

# ---- cmd prompt decoration ---- #
EC_SYS_ID="TutorialTest"
EC_SYS_TYPE="FUNC_TEST"
export EC_SYS_ID EC_SYS_TYPE
PS1='[${EC_SYS_ID}(${EC_SYS_TYPE})][\u@\h:\w]\$ '
export PS1

# ---- ec environment ---- #
EC_QSL_PATH=${EC_SYS_PATH}/bin/ec/libraries/qsl/
EC_ETC_PATH=${EC_SYS_PATH}/bin/ec/test/tutorialTest/etc/
export EC_QSL_PATH EC_ETC_PATH

# ---- yak environment ---- #
YAK_PATH=$EC_SYS_PATH/bin/yak/
YAK_OPTS="-c $EC_ETC_PATH/system.cfg -v less -s $EC_SYS_PATH/data/test/${EC_SYS_ID}/yak/yak.status -l $EC_SYS_PATH/log/test/${EC_SYS_ID}/yak/yak.log"
export YAK_OPTS
. $YAK_PATH/yak_complete_bash.sh
NEWPATH=$YAK_PATH

# ---- q environment ---- #
# search for q binary in bin/q/ (cover MacOS and Linux)
if [ "$(uname)" == "Darwin" ]; then OS_LETTER=m; else OS_LETTER=l; fi
if [ -f "$EC_SYS_PATH/bin/q/${OS_LETTER}32/q" ]; then QHOME=$EC_SYS_PATH/bin/q/ && NEWPATH=$QHOME/${OS_LETTER}32/:$NEWPATH; fi
if [ -f "$EC_SYS_PATH/bin/q/${OS_LETTER}64/q" ]; then QHOME=$EC_SYS_PATH/bin/q/ && NEWPATH=$QHOME/${OS_LETTER}64/:$NEWPATH; fi
# QLIC set to etc/, in case of q64, k4.lic should be placed in $QLIC directory
QLIC=$EC_SYS_PATH/etc/
export QHOME QLIC

# ---- path ---- #
export OLDPATH=${OLDPATH:-$PATH}
PATH=$NEWPATH:$OLDPATH
export PATH
#check if q is available on PATH
which q >/dev/null || echo "Missing q binary. Please install q in $EC_SYS_PATH/bin/q/ directory, see ec/tutorial/Installation.md for instructions."

echo env for system ${EC_SYS_ID}/${EC_SYS_TYPE} loaded


components/hdb/test/etc/env.sh
if [ ! -d "bin" ]; then
  echo "Expecting bin subdirectory" && return
fi
export EC_SYS_PATH=${PWD}

# ---- cmd prompt decoration ---- #
EC_SYS_ID="hdbTest"
EC_SYS_TYPE="FUNC_TEST"
export EC_SYS_ID EC_SYS_TYPE
PS1='[${EC_SYS_ID}(${EC_SYS_TYPE})][\u@\h:\w]\$ '
export PS1

# ---- ec environment ---- #
EC_QSL_PATH=${EC_SYS_PATH}/bin/ec/libraries/qsl/
EC_ETC_PATH=${EC_SYS_PATH}/bin/ec/components/hdb/test/etc/
export EC_QSL_PATH EC_ETC_PATH

# ---- yak environment ---- #
YAK_PATH=$EC_SYS_PATH/bin/yak/
YAK_OPTS="-c $EC_ETC_PATH/system.cfg -v less -s $EC_SYS_PATH/data/test/${EC_SYS_ID}/yak/yak.status -l $EC_SYS_PATH/log/test/${EC_SYS_ID}/yak/yak.log"
export YAK_OPTS
. $YAK_PATH/yak_complete_bash.sh
NEWPATH=$YAK_PATH

# ---- q environment ---- #
# search for q binary in bin/q/ (cover MacOS and Linux)
if [ "$(uname)" == "Darwin" ]; then OS_LETTER=m; else OS_LETTER=l; fi
if [ -f "$EC_SYS_PATH/bin/q/${OS_LETTER}32/q" ]; then QHOME=$EC_SYS_PATH/bin/q/ && NEWPATH=$QHOME/${OS_LETTER}32/:$NEWPATH; fi
if [ -f "$EC_SYS_PATH/bin/q/${OS_LETTER}64/q" ]; then QHOME=$EC_SYS_PATH/bin/q/ && NEWPATH=$QHOME/${OS_LETTER}64/:$NEWPATH; fi
# QLIC set to etc/, in case of q64, k4.lic should be placed in $QLIC directory
QLIC=$EC_SYS_PATH/etc/
export QHOME QLIC

# ---- path ---- #
export OLDPATH=${OLDPATH:-$PATH}
PATH=$NEWPATH:$OLDPATH
export PATH
#check if q is available on PATH
which q >/dev/null || echo "Missing q binary. Please install q in $EC_SYS_PATH/bin/q/ directory, see ec/tutorial/Installation.md for instructions."

echo env for system ${EC_SYS_ID}/${EC_SYS_TYPE} loaded


tutorial/Lesson02/etc/env.sh
if [ ! -d "bin" ]; then
  echo "Expecting bin subdirectory" && return
fi
export EC_SYS_PATH=${PWD}

# ---- cmd prompt decoration ---- #
EC_SYS_ID="DemoSystem"
EC_SYS_TYPE="Lesson02"
export EC_SYS_ID EC_SYS_TYPE
PS1='[${EC_SYS_ID}(${EC_SYS_TYPE})][\u@\h:\w]\$ '
export PS1

# ---- ec environment ---- #
EC_QSL_PATH=${EC_SYS_PATH}/bin/ec/libraries/qsl/
EC_ETC_PATH=${EC_SYS_PATH}/etc
export EC_QSL_PATH EC_ETC_PATH

# ---- yak environment ---- #
YAK_PATH=$EC_SYS_PATH/bin/yak/
YAK_OPTS="-c $EC_ETC_PATH/system.cfg -v less -s $EC_SYS_PATH/data/test/${EC_SYS_ID}/yak/yak.status -l $EC_SYS_PATH/log/test/${EC_SYS_ID}/yak/yak.log"
export YAK_OPTS
. $YAK_PATH/yak_complete_bash.sh
NEWPATH=$YAK_PATH

# ---- q environment ---- #
# search for q binary in bin/q/ (cover MacOS and Linux)
if [ "$(uname)" == "Darwin" ]; then OS_LETTER=m; else OS_LETTER=l; fi
if [ -f "$EC_SYS_PATH/bin/q/${OS_LETTER}32/q" ]; then QHOME=$EC_SYS_PATH/bin/q/ && NEWPATH=$QHOME/${OS_LETTER}32/:$NEWPATH; fi
if [ -f "$EC_SYS_PATH/bin/q/${OS_LETTER}64/q" ]; then QHOME=$EC_SYS_PATH/bin/q/ && NEWPATH=$QHOME/${OS_LETTER}64/:$NEWPATH; fi
# QLIC set to etc/, in case of q64, k4.lic should be placed in $QLIC directory
QLIC=$EC_SYS_PATH/etc/
export QHOME QLIC

# ---- path ---- #
export OLDPATH=${OLDPATH:-$PATH}
PATH=$NEWPATH:$OLDPATH
export PATH
#check if q is available on PATH
which q >/dev/null || echo "Missing q binary. Please install q in $EC_SYS_PATH/bin/q/ directory, see ec/tutorial/Installation.md for instructions."

echo env for system ${EC_SYS_ID}/${EC_SYS_TYPE} loaded

